magma/circuit.py

```diff
@@ -232,12 +231,11 @@ def _add_intermediate_value(value):
     `CircuitKind.__repr__`.
     """
     root = value.name.root()
-    # root is either None, or a NamedRef. If it is a NamedRef which refers to a
-    # constant, then we can proceed with the empty set. Otherwise, we seed the
-    # set with the root value.
-    if root is None or root.value().const():
-        return OrderedIdentitySet()
-    return OrderedIdentitySet((root.value(),))
+    # If it is a NamedRef which doesn't refer to a constant, then we seed the
+    # set with the root value. Otherwise, we can proceed with the empty set
+    if type(root) is NamedRef and not root.value().const():
```

Comment:
should this be `isinstance(root, NamedRef)`? this seems brittle.

If it's this way because `InstRef` and `DefnRef` are subclasses of `NamedRef` but we don't want to include them, then there should be a more well-defined way of dealing with these.

I think this is fine for now, but we should in the future clean this up. Just dropping this here for now to document: every instance of `Ref` should have 3 attributes: named, bound, derived. They should be assigned as follows:
* Everything besides AnonRef has named=True; AnonRef has named=False.
* (Lazy)InstRef and (Lazy)DefnRef should all have bound=True. Their "derived" refs (e.g. array slices or tuple getitems) should also have bound=True.
* Only ArrayRef and TupleRef should have derived=True. All others should have derived=False.

I think this covers all our cases and avoids a lot of the ad-hoc logic floating around regarding refs!

(Again, fine to leave as is for now. Probably will this roll this into the backend change).

Reply:
> If it's this way because InstRef and DefnRef are subclasses of NamedRef but we don't want to include them, then there should be a more well-defined way of dealing with these
This is why

I think we can add a `TempRef` that is a `NamedRef` subclass, but not in the tree of Inst/Defn Ref to resolve this.